drivebrain_core_base/include/public/MsgLogger.hpp

```diff
@@ -0,0 +1,148 @@
+#ifndef __MSG_LOGGER__
+#define __MSG_LOGGER__
+
+#include <memory>
+#include <string>
+
+#include <deque>
+#include <thread>
+#include <functional>
+#include <condition_variable>
+#include <mutex>
+
+#include "hytech_msgs.pb.h"
+#include <google/protobuf/message.h>
+
+namespace core
+{
+    template <typename MsgType>
+    class MsgLogger
+    {
+    public:
+        struct ThreadSafeOutput
+        {
+            std::mutex mtx;
+            std::condition_variable cv;
+            std::deque<MsgType> deque;
+        };
+
+        MsgLogger(const std::string &log_file_extension,
+                  bool init_logging,
+                  std::function<void(MsgType)> logger_msg_func,
+                  std::function<void()> stop_log_func,
+                  std::function<void(const std::string &)> open_log_func,
+                  std::function<void(MsgType)> live_msg_output_func) : _logger_msg_function(logger_msg_func),
+                                                                       _live_msg_output_func(live_msg_output_func),
+                                                                       _log_file_extension(log_file_extension),
+                                                                       _stop_log_function(stop_log_func),
+                                                                       _open_log_function(open_log_func)
+        {
+            _running = true;
+
+            if (init_logging)
+            {
+                auto log_name = _generate_log_name(_log_file_extension);
+                _current_log_name = log_name;
+                _open_log_function(log_name);
+                _logging = init_logging;
+            }
+
+        }
+
+        ~MsgLogger()
+        {
+            // should be fine for the most part to not lock here, this is the only place we are writing these bools
+            _running = false;
+            _logging = false;
+            _stop_log_function();
+            std::cout << "safely exited" << std::endl;
+        }
+
+        void log_msg(MsgType msg)
+        {
+            // TODO maybe make this also more thread safe ... ?
```

Comment:
Might want to add mutex here bc I think there could be situations where log_msg is being called by multiple parts of the code at once.